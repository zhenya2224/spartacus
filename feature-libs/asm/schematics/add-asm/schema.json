{
  "$schema": "http://json-schema.org/schema",
  "$id": "AsmSchematics",
  "title": "ASM Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the asm features.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "default": ["ASM"]
    }
  },
  "required": []
}
