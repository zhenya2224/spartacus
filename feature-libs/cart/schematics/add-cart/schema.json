{
  "$schema": "http://json-schema.org/schema",
  "$id": "CartSchematics",
  "title": "Cart Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the Saved cart feature.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": ["Saved-Cart", "Quick-Order", "Import-Export"],
        "type": "string"
      },
      "default": ["Saved-Cart", "Quick-Order", "Import-Export"],
      "x-prompt": {
        "message": "Which Cart features would you like to set up?",
        "type": "list",
        "items": [
          {
            "value": "Saved-Cart",
            "label": "Saved Cart"
          },
          {
            "value": "Quick-Order",
            "label": "Quick Order"
          },
          {
            "value": "Import-Export",
            "label": "Import/Export"
          }
        ]
      }
    }
  },
  "required": []
}
