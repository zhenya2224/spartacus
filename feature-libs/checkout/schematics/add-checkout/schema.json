{
  "$schema": "http://json-schema.org/schema",
  "$id": "CheckoutSchematics",
  "title": "Checkout Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the Checkout feature.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "default": ["Checkout"]
    }
  },
  "required": []
}
