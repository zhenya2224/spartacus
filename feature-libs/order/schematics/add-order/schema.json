{
  "$schema": "http://json-schema.org/schema",
  "$id": "OrderSchematics",
  "title": "Order Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the Order feature.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "default": ["Order"]
    }
  },
  "required": []
}
