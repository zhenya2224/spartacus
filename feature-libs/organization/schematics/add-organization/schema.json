{
  "$schema": "http://json-schema.org/schema",
  "$id": "OrganizationSchematics",
  "title": "Organization Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the organization features.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": ["Administration", "Order-Approval"],
        "type": "string"
      },
      "default": ["Administration", "Order-Approval"],
      "x-prompt": {
        "message": "Which Organization features would you like to set up?",
        "type": "list",
        "items": [
          {
            "value": "Administration",
            "label": "Administration (b2b feature)"
          },
          { "value": "Order-Approval", "label": "Order Approval (b2b feature)" }
        ]
      }
    }
  },
  "required": []
}
