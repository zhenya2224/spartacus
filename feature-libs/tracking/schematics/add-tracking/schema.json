{
  "$schema": "http://json-schema.org/schema",
  "$id": "TrackingSchematics",
  "title": "Tracking Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the tracking features.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": ["Personalization", "TMS-GTM", "TMS-AEPL"],
        "type": "string"
      },
      "default": ["Personalization"],
      "x-prompt": {
        "message": "Which Tracking features would you like to set up?",
        "type": "list",
        "items": [
          { "value": "Personalization", "label": "Personalization" },
          {
            "value": "TMS-GTM",
            "label": "Tag Management System - Google Tag Manager"
          },
          {
            "value": "TMS-AEPL",
            "label": "Tag Management System - Adobe Experience Platform Launch"
          }
        ]
      }
    }
  },
  "required": []
}
