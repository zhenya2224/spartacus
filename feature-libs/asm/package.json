{
  "name": "@spartacus/asm",
  "version": "4.1.0-next.0",
  "description": "ASM feature library for Spartacus",
  "keywords": [
    "spartacus",
    "framework",
    "storefront",
    "asm"
  ],
  "homepage": "https://github.com/SAP/spartacus",
  "repository": "https://github.com/SAP/spartacus/tree/develop/feature-libs/asm",
  "license": "Apache-2.0",
  "scripts": {
    "build:schematics": "yarn clean:schematics && ../../node_modules/.bin/tsc -p ./tsconfig.schematics.json",
    "clean:schematics": "../../node_modules/.bin/rimraf \"schematics/**/*.js\" \"schematics/**/*.js.map\" \"schematics/**/*.d.ts\"",
    "test:schematics": "yarn --cwd ../../projects/schematics/ run clean && yarn clean:schematics && ../../node_modules/.bin/jest --config ./jest.schematics.config.js"
  },
  "dependencies": {
    "tslib": "^2.3.0"
  },
  "peerDependencies": {
    "@angular-devkit/schematics": "^12.0.5",
    "@angular/common": "^12.0.5",
    "@angular/core": "^12.0.5",
    "@angular/forms": "^12.0.5",
    "@ngrx/effects": "^12.1.0",
    "@ngrx/store": "^12.1.0",
    "@spartacus/core": "4.1.0-next.0",
    "@spartacus/schematics": "4.1.0-next.0",
    "@spartacus/storefront": "4.1.0-next.0",
    "rxjs": "^6.6.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "schematics": "./schematics/collection.json"
}
