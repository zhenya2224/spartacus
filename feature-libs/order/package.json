{
  "name": "@spartacus/order",
  "version": "4.2.0",
  "description": "Order feature library for Spartacus",
  "keywords": [
    "spartacus",
    "framework",
    "storefront",
    "order"
  ],
  "homepage": "https://github.com/SAP/spartacus",
  "repository": "https://github.com/SAP/spartacus/tree/develop/feature-libs/order",
  "license": "Apache-2.0",
  "scripts": {
    "build:schematics": "yarn clean:schematics && ../../node_modules/.bin/tsc -p ./tsconfig.schematics.json",
    "clean:schematics": "../../node_modules/.bin/rimraf \"schematics/**/*.js\" \"schematics/**/*.js.map\" \"schematics/**/*.d.ts\"",
    "test:schematics": "yarn --cwd ../../projects/schematics/ run clean && yarn clean:schematics && ../../node_modules/.bin/jest --config ./jest.schematics.config.js"
  },
  "dependencies": {
    "tslib": "^2.3.0"
  },
  "peerDependencies": {
    "@angular-devkit/schematics": "^12.0.5",
    "@angular/common": "^12.0.5",
    "@angular/core": "^12.0.5",
    "@angular/forms": "^12.0.5",
    "@angular/router": "^12.0.5",
    "@ng-bootstrap/ng-bootstrap": "^10.0.0",
    "@ng-select/ng-select": "^7.0.1",
    "@ngrx/effects": "^12.1.0",
    "@ngrx/store": "^12.1.0",
    "@spartacus/core": "4.1.0-next.0",
    "@spartacus/schematics": "4.1.0-next.0",
    "@spartacus/storefront": "4.1.0-next.0",
    "@spartacus/styles": "4.1.0-next.0",
    "rxjs": "^6.6.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "schematics": "./schematics/collection.json"
}
