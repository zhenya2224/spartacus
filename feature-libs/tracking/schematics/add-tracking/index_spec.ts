/// <reference types="jest" />

import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import {
  CLI_TRACKING_PERSONALIZATION_FEATURE,
  CLI_TRACKING_TMS_AEP_FEATURE,
  CLI_TRACKING_TMS_GTM_FEATURE,
  LibraryOptions as SpartacusTrackingOptions,
  SpartacusOptions,
  SPARTACUS_SCHEMATICS,
} from '@spartacus/schematics';
import * as path from 'path';
import { peerDependencies } from '../../package.json';

const collectionPath = path.join(__dirname, '../collection.json');
const personalizationModulePath =
  'src/app/spartacus/features/tracking/personalization-feature.module.ts';
const tagManagementModulePath =
  'src/app/spartacus/features/tracking/tag-management-feature.module.ts';

describe('Spartacus Tracking schematics: ng-add', () => {
  const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

  let appTree: UnitTestTree;

  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };

  const spartacusDefaultOptions: SpartacusOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [],
  };

  const libraryNoFeaturesOptions: SpartacusTrackingOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [],
  };

  const personalizationFeatureOptions: SpartacusTrackingOptions = {
    ...libraryNoFeaturesOptions,
    features: [CLI_TRACKING_PERSONALIZATION_FEATURE],
  };

  const gtmFeatureOptions: SpartacusTrackingOptions = {
    ...libraryNoFeaturesOptions,
    features: [CLI_TRACKING_TMS_GTM_FEATURE],
  };

  const aepFeatureOptions: SpartacusTrackingOptions = {
    ...libraryNoFeaturesOptions,
    features: [CLI_TRACKING_TMS_AEP_FEATURE],
  };

  beforeEach(async () => {
    schematicRunner.registerCollection(
      SPARTACUS_SCHEMATICS,
      '../../projects/schematics/src/collection.json'
    );

    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        SPARTACUS_SCHEMATICS,
        'ng-add',
        { ...spartacusDefaultOptions, name: 'schematics-test' },
        appTree
      )
      .toPromise();
  });

  describe('Without features', () => {
    beforeEach(async () => {
      appTree = await schematicRunner
        .runSchematicAsync('ng-add', libraryNoFeaturesOptions, appTree)
        .toPromise();
    });

    it('should not create any of the feature modules', () => {
      expect(appTree.exists(personalizationModulePath)).toBeFalsy();
      expect(appTree.exists(tagManagementModulePath)).toBeFalsy();
    });

    it('should install necessary Spartacus libraries', () => {
      const packageJson = JSON.parse(appTree.readContent('package.json'));
      let dependencies: Record<string, string> = {};
      dependencies = { ...packageJson.dependencies };
      dependencies = { ...dependencies, ...packageJson.devDependencies };

      for (const toAdd in peerDependencies) {
        // skip the SPARTACUS_SCHEMATICS, as those are added only when running by the Angular CLI, and not in the testing environment
        if (
          !peerDependencies.hasOwnProperty(toAdd) ||
          toAdd === SPARTACUS_SCHEMATICS
        ) {
          continue;
        }
        // TODO: after 4.0: use this test, as we'll have synced versions between lib's and root package.json
        // const expectedVersion = (peerDependencies as Record<
        //   string,
        //   string
        // >)[toAdd];
        const expectedDependency = dependencies[toAdd];
        expect(expectedDependency).toBeTruthy();
        // expect(expectedDependency).toEqual(expectedVersion);
      }
    });
  });

  describe('Personalization feature', () => {
    describe('general setup', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync('ng-add', personalizationFeatureOptions, appTree)
          .toPromise();
      });

      it('should add the feature using the lazy loading syntax', async () => {
        const personalizationModule = appTree.readContent(
          personalizationModulePath
        );
        expect(personalizationModule).toMatchSnapshot();
      });
    });

    describe('eager loading', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync(
            'ng-add',
            {
              ...personalizationFeatureOptions,
              lazy: false,
            },
            appTree
          )
          .toPromise();
      });

      it('should import appropriate modules', async () => {
        const module = appTree.readContent(personalizationModulePath);
        expect(module).toMatchSnapshot();
      });
    });
  });

  describe('Tag Management feature', () => {
    describe('GTM', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync('ng-add', gtmFeatureOptions, appTree)
          .toPromise();
      });
      describe('general setup', () => {
        it('should import appropriate modules (without lazy loaded syntax)', async () => {
          const tagManagementModule = appTree.readContent(
            tagManagementModulePath
          );
          expect(tagManagementModule).toMatchSnapshot();
        });
      });
    });

    describe('AEP', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync('ng-add', aepFeatureOptions, appTree)
          .toPromise();
      });

      describe('general setup', () => {
        it('should import appropriate modules (without lazy loaded syntax)', async () => {
          const tagManagementModule = appTree.readContent(
            tagManagementModulePath
          );
          expect(tagManagementModule).toMatchSnapshot();
        });
      });
    });

    describe('GTM and AEP', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync(
            'ng-add',
            {
              ...libraryNoFeaturesOptions,
              features: [
                CLI_TRACKING_TMS_GTM_FEATURE,
                CLI_TRACKING_TMS_AEP_FEATURE,
              ],
            },
            appTree
          )
          .toPromise();
      });

      describe('general setup', () => {
        it('should import appropriate modules (without lazy loaded syntax)', async () => {
          const tagManagementModule = appTree.readContent(
            tagManagementModulePath
          );
          expect(tagManagementModule).toMatchSnapshot();
        });
      });
    });
  });
});
