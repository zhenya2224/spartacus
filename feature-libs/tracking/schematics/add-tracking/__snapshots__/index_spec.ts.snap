// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spartacus Tracking schematics: ng-add Personalization feature eager loading should import appropriate modules 1`] = `
"import { NgModule } from '@angular/core';
import { PersonalizationModule } from \\"@spartacus/tracking/personalization\\";
import { PersonalizationRootModule } from \\"@spartacus/tracking/personalization/root\\";

@NgModule({
  declarations: [],
  imports: [
    PersonalizationRootModule,
    PersonalizationModule
  ]
})
export class PersonalizationFeatureModule { }
"
`;

exports[`Spartacus Tracking schematics: ng-add Personalization feature general setup should add the feature using the lazy loading syntax 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, provideConfig } from \\"@spartacus/core\\";
import { PersonalizationRootModule, PERSONALIZATION_FEATURE } from \\"@spartacus/tracking/personalization/root\\";

@NgModule({
  declarations: [],
  imports: [
    PersonalizationRootModule
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      [PERSONALIZATION_FEATURE]: {
        module: () =>
          import('@spartacus/tracking/personalization').then((m) => m.PersonalizationModule),
      },
    }
  })]
})
export class PersonalizationFeatureModule { }
"
`;

exports[`Spartacus Tracking schematics: ng-add Tag Management feature AEP general setup should import appropriate modules (without lazy loaded syntax) 1`] = `
"import { NgModule } from '@angular/core';
import { provideConfig } from \\"@spartacus/core\\";
import { AepModule } from \\"@spartacus/tracking/tms/aep\\";
import { BaseTmsModule, TmsConfig } from \\"@spartacus/tracking/tms/core\\";

@NgModule({
  declarations: [],
  imports: [
    BaseTmsModule.forRoot(),
    AepModule
  ],
  providers: [provideConfig(<TmsConfig>{
    tagManager: {
      aep: {
        events: [],
      },
    },
  })]
})
export class TagManagementFeatureModule { }
"
`;

exports[`Spartacus Tracking schematics: ng-add Tag Management feature GTM and AEP general setup should import appropriate modules (without lazy loaded syntax) 1`] = `
"import { NgModule } from '@angular/core';
import { provideConfig } from \\"@spartacus/core\\";
import { AepModule } from \\"@spartacus/tracking/tms/aep\\";
import { BaseTmsModule, TmsConfig } from \\"@spartacus/tracking/tms/core\\";
import { GtmModule } from \\"@spartacus/tracking/tms/gtm\\";

@NgModule({
  declarations: [],
  imports: [
    BaseTmsModule.forRoot(),
    GtmModule,
    AepModule
  ],
  providers: [provideConfig(<TmsConfig>{
    tagManager: {
      gtm: {
        events: [],
      },
    },
  }),
  provideConfig(<TmsConfig>{
    tagManager: {
      aep: {
        events: [],
      },
    },
  })
  ]
})
export class TagManagementFeatureModule { }
"
`;

exports[`Spartacus Tracking schematics: ng-add Tag Management feature GTM general setup should import appropriate modules (without lazy loaded syntax) 1`] = `
"import { NgModule } from '@angular/core';
import { provideConfig } from \\"@spartacus/core\\";
import { BaseTmsModule, TmsConfig } from \\"@spartacus/tracking/tms/core\\";
import { GtmModule } from \\"@spartacus/tracking/tms/gtm\\";

@NgModule({
  declarations: [],
  imports: [
    BaseTmsModule.forRoot(),
    GtmModule
  ],
  providers: [provideConfig(<TmsConfig>{
    tagManager: {
      gtm: {
        events: [],
      },
    },
  })]
})
export class TagManagementFeatureModule { }
"
`;
